async/cohttp_async.ml

```diff
@@ -156,25 +156,60 @@ end
 
 module Client = struct
 
+  let read_request ic =
+    Response.read ic >>| function
+    | `Eof -> failwith "Connection closed by remote host"
+    | `Invalid reason -> failwith reason
```

Comment:
I think that defining an `Invalid_response` exception would be fine.

But this PR doesn't introduce any kind of new error handling. This function was simply extracted out of request to be reusable in `callv`.

However, you remind of a good point. I think that both async's and lwt's callv shouldn't raise at all because it might be possible that raising on a bad response could possibly make it impossible for the user to retrieve the previous good responses (although the connection should be terminated regardless)